waspc/data/Generator/templates/react-app/vite/detectServerImports.ts

```
@@ -0,0 +1,66 @@
+import { type Plugin } from "vite";
+import path from "path";
+
+export function detectServerImports(): Plugin {
+  return {
+    name: 'wasp-detect-server-imports',
+    transform(code, filePath) {
+      const isInDotWaspFolder = filePath.includes("/.wasp/");
+
+      // We don't want to check for server imports in the Wasp 
+      // framework code.
+      if (isInDotWaspFolder) {
+        return;
+      }
+
+      const imports = getImportsFromCode(code);
+
+      for (const imp of imports) {
+        if (imp.moduleName.startsWith("wasp/server")) {
+          throw new Error(getServerImportErrorMessage(imp, filePath));
+        }
+      }
+    },
+  };
+}
+
+type Import = {
+  importStatement: string;
+  moduleName: string;
+}
+
+const importStatementRegex = /\s*import\s+(?:(?:[\w*\s{},]*)\s+from\s+)?(['"`])([^'"`]+)\1\s*/g;
```

I didn't dive deep, just wanted to check, since this is regex, is there a chance of false positives? If so, in which situations, and are we ok with that? If not, can we do it better?

---

Vite [doesn't expose](https://github.com/vitejs/vite/issues/13486#issuecomment-1598188384) AST information to the plugins. One the popular plugins [unplugin-ast](https://github.com/unplugin/unplugin-ast) uses the [Babel](https://www.npmjs.com/package/ast-kit) parser under the hood to parse files into AST.

I didn't want to go down this route due to potential performance issues, but I have to admit that I haven't timed anything. I just assumed that regexes will be good enough.

I went with a good enough approach because we are not catch all server imports anyways. We are only focusing on `wasp/server/*` modules (for which we know they shouldn't be imported on the client). But... we don't catch stuff like `some-node-dep`.

I will explore the AST approach and see how it works and report back. I'll also explore the JS imports syntax and see if our regex matches all the cases.

---

It turns out we don't need to parse or use regexes at all. @sodic pointed me in the right direction, there is a different plugin hook called `resolveId` which is called when Vite wants to resolve an import (so it already parsed everything) and we can decide if we want to raise a meaningful error if we notice it's a `wasp/server/*` import :)